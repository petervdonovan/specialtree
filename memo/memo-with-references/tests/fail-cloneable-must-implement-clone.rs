use memo_with_references::memo;

// This should fail because NoClone doesn't implement Clone
// but it's a cloneable argument (not a reference with cache lifetime)
struct NoClone;

#[memo('a)]
fn test_cloneable_must_implement_clone<'a, 'b>(
    data: &'a String,  // frozen (matches cache lifetime)
    other: &'b String, // cloneable (different lifetime) - should require Clone
    owned: NoClone,    // cloneable (owned) - should require Clone
) -> &'a String {
    data
}

fn main() {}
