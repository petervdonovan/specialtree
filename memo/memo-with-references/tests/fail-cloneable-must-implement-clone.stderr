error[E0277]: the trait bound `NoClone: Hash` is not satisfied
 --> tests/fail-cloneable-must-implement-clone.rs:7:1
  |
7 | #[memo('a)]
  | ^^^^^^^^^^^ the trait `Hash` is not implemented for `NoClone`
  |
  = help: see issue #48214
  = note: this error originates in the attribute macro `memo` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `NoClone` with `#[derive(Hash)]`
  |
5 + #[derive(Hash)]
6 | struct NoClone;
  |
help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
  |
1 + #![feature(trivial_bounds)]
  |

error[E0277]: the trait bound `NoClone: Eq` is not satisfied
 --> tests/fail-cloneable-must-implement-clone.rs:7:1
  |
7 | #[memo('a)]
  | ^^^^^^^^^^^ the trait `Eq` is not implemented for `NoClone`
  |
  = help: see issue #48214
  = note: this error originates in the attribute macro `memo` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `NoClone` with `#[derive(Eq)]`
  |
5 + #[derive(Eq)]
6 | struct NoClone;
  |
help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
  |
1 + #![feature(trivial_bounds)]
  |

error[E0277]: the trait bound `NoClone: Clone` is not satisfied
 --> tests/fail-cloneable-must-implement-clone.rs:7:1
  |
7 | #[memo('a)]
  | ^^^^^^^^^^^ the trait `Clone` is not implemented for `NoClone`
  |
  = help: see issue #48214
  = note: this error originates in the attribute macro `memo` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `NoClone` with `#[derive(Clone)]`
  |
5 + #[derive(Clone)]
6 | struct NoClone;
  |
help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
  |
1 + #![feature(trivial_bounds)]
  |
